### ISSUE TITLE
 About the Learning  Performance 

### ISSUE THREAD

**Issue by JieMEI1994** [JulesVerny/PongReinforcementLearning#1]
Thanks for your code and description.
It help a lot to understand the DQN algorithm, but my code can not gain the learning performance as good as yours. I wonder whether there is some tricks you use? Could you please share your experience as well?

Here is my code: https://github.com/JieMEI1994/HelloWorld

Thank you so much.

**Comment by JulesVerny** [comment]
Jie ?

 

Your code is rather more sophisticated and generalised than mine, so it would be difficult to appreciate any errors.  I don’t see anything obviously wrong in your  core DQN code.   I note that your DQN keras model is more layers than mine with dropout etc .  Its only a 4 action => Q Value estimate, so I would not have thought we need a very complicated Neural net Regression model.  Starting simple is usually better. Does your code show any performance improvement at all ? 

 

Have you run my code, and repeated my results ? 

I am still new to all this DQN, so I typically only make small changes to other peoples running code, (c.f. my ref to https://jaromiru.com/2016/10/03/lets-make-a-dqn-implementation/)  to make sure I can repeat it, before doing some experiments on adjusting some DQN parameters and small changes, before I attempt any significant generalisation rewrite.   

 

Sorry I cannot be more Helpful

 

Jules

**Comment by JieMEI1994** [comment]
Yes, please just call me Jie.

Actually I have test your code and get a very good performance, it takes about 5 min to learn how to play the Pong game as the increasing total score shown.
But when I tried to write my own code based on yours, I still could not get the result as well as yours. I  found out you didn't use two neural network structures but two different q value, which means you update the q value online. So I wonder whether you mean to do that and why?
And Thanks, my network is too complicated and I will try a simple one afterward.

I will update my github with some result figure, please check. I appreciate to be able to discuss with you.
Thanks for your kind reply again.

Jie
________________________________

**Comment by JieMEI1994** [comment]
I have update my GitHub, you can check my learning performance. But considering the different evaluation method, it will not look the same as yours.
And thanks for your recommended material of the DQN. I noticed that the reference materials used by that blog are based on the 2013 version which does not use target network structure. And I have not idea whether it matter or not. But I have found out that shallow neural structure helps improve the learning performance.
I also found out that we both didn't use state = next_state, I wonder the reason. Do you have any idea?

**Comment by JulesVerny** [comment]
Jie,
Good to hear you are getting better results. Remember the Q Function in this case is pretty simple. Only about 5 state features into a Q Function.  There is both the risk or overfitting, longer convergence or no convergence at all with ever more complex DQN nets.  So its a difficult balance how sophisticated to develop the DQN Nueral net model against your Input/output shape. I think there is an ML maxime, start simple, and only make more complicated if you feel that your model is not performing, underfitting.  Remember there is in fact no mathematical guarantee that a Nueral net DQN algorithm will ever converge. (as apposed to tabular Q Table which will always converge.)
The recommendation for  DQN neural  nets is that the same Q Function approximator can estimate a Q value for each of the (Action)outputs  [So Three output Y values against the 5 inputs] No need for a separate NN for each Action, if that is what you  are suggesting. 

I am now trying a Convolutional network, based upon taking the whole game image  [ScreenImage = pygame.surfarray.array3d(pygame.display.get_surface())]  and reducing this down to a 40x40 pixel surface into a convolutional net.  This takes much longer about 250,000 epochs (or a couple of days) to get any where near reasonable game convergence, and still not as good a player as the feature based.  I will maybe keep tinkering, but is a slow frustrating experience. (Even on my GTX 1080 Ti GPU)  

Cheers and Good luck

Julian

**Comment by JulesVerny** [comment]
Jie

 

Good to hear you are getting better results. Remember the Q Function in this case is pretty simple. Only about 5 state features into a Q Function.  There is both the risk or overfitting, longer convergence or no convergence at all with ever more complex DQN nets.  So its a difficult balance how sophisticated to develop the DQN Nueral net model against your Input/output shape. I think there is an ML maxime, start simple, and only make more complicated if you feel that your model is not performing, underfitting.  Remember there is in fact no mathematical guarantee that a Nueral net DQN algorithm will ever converge. (as apposed to tabular Q Table which will always converge.)

The recommendation for  DQN neural  nets is that the same Q Function approximator can estimate a Q value for each of the (Action)outputs  [So Three output Y values against the 5 inputs] No need for a separate NN for each Action, if that is what you  are suggesting. 

 

I am now trying a Convolutional network, based upon taking the whole game image  [ScreenImage = pygame.surfarray.array3d(pygame.display.get_surface())]  and reducing this down to a 40x40 pixel surface into a convolutional net.  This takes much longer about 250,000 epochs (or a couple of days) to get any where near reasonable game convergence, and still not as good a player as the feature based.  I will maybe keep tinkering, but is a slow frustrating experience. (Even on my GTX 1080 Ti GPU)  

 

I have put the above as comment on your GitHub

 

Cheers and Good luck

 

Julian

**Comment by JieMEI1994** [comment]
Julian,

yes, thanks for your advice. I have fixed the problem, you can check my GitHub where shows the 2015 version of DQN. I hope I can also help you somehow.

I plan to implement policy gradient  based method with the same environment.

talk to you later if I have a good result.

Jie
________________________________
